keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -775,129 +788,188 @@ Some flow schemata.
+  - rule: # Delegated auth[n/z] requests
+      verbs: ['*']
+      apiGroups: ['authentication.k8s.io', 'authorization.k8s.io']
+      resources: ['*']
+    subjects:
+    - kind: Group
+      name: "system:authenticated"     
+    - kind: Group
+      name: "system:unauthenticated"     
 ```
 
 ```yaml
 kind: FlowSchema
 meta:
-  name: system-high
+  name: nodes
 spec:
-  requestPriority:
+  matchingPrecedence: 2500
+  priorityLevelConfiguration:
     name: system-high
-  flowDistinguisher:
-    source: user
-    # no transformation in this case
-  match:
-  - and: # heartbeats by nodes
-    - superSet:
-      field: groups
-      set: [ "system:nodes" ]
-    - equals:
-      field: resource
-      value: nodes
-  - and: # kubelet and kube-proxy ops on system objects
-    - superSet:
-      field: groups
-      set: [ "system:nodes" ]
-    - equals:
-      field: namespace
-      value: kube-system
-  - and: # leader elections for system controllers
-    - patternMatch:
-      field: user
-      pattern: system:controller:.*
-    - inSet:
-      field: resource
-      set: [ "endpoints", "configmaps", "leases" ]
-    - equals:
-      field: namespace
-      value: kube-system
+  distinguisherMethod:
+    type: ByUser
+  rules:  
+  - rule: # All resource requests from node group
+      verbs: ['*']
+      apiGroups: ['*']
+      resources: ['*']
+    subjects:
+    - kind: Group
+      name: "system:nodes"
+  - rule: # All non-resource requests from node group
+      verbs: ['*']
+      nonResourceURLs: ['*']
+    subjects:
+    - kind: Group
+      name: "system:nodes"    
 ```
 
 ```yaml
 kind: FlowSchema
 meta:
-  name: system-low
+  name: leader-election
 spec:
-  matchingPriority: 900
-  requestPriority:
-    name: system-low
-  flowDistinguisher:
-    source: user
-    # no transformation in this case
-  match:
-  - and: # the garbage collector
-    - equals:
-      field: user
-      value: system:controller:garbage-collector
+  matchingPrecedence: 2500
+  priorityLevelConfiguration:
+    name: system-high
+  distinguisherMethod:
+    type: ByUser
+  rules:  
+  - rule: # Leader-election requests from kube-controller-manager/kube-scheduler
+      verbs: ['*']
+      apiGroups: ['coordination.k8s.io']
+      resources: ['leases']
+    subjects:
+    - kind: User
+      name: "system:kube-controller-manager"    
+    - kind: User
+      name: "system:kube-scheduler" 
 ```
 
 ```yaml
 kind: FlowSchema
 meta:
-  name: workload-high
+  name: kube-manager-controller
 spec:
-  requestPriority:
-    name: workload-high
-  flowDistinguisher:
-    source: namespace
-    # no transformation in this case
-  match:
-  - and: # users using kubectl
-    - notPatternMatch:
-      field: user
-      pattern: system:serviceaccount:.*
+  matchingPrecedence: 3500
+  priorityLevelConfiguration:
+    name: system-low
+  distinguisherMethod:
+    type: ByUser
+  rules:  
+  - rule: # All resource requests from kube-controller-manager
```

Comment:
In the KEP currently, almost all of these are given much lower priority.  They are part of "workload-low", while kubectl requests from users are prioritized more highly at "workload-high".  I think moving controller requests above user kubectl requests is not an improvement.

Reply:
i think it's difficult to wirte a rule for "kubectl requests".. kubectl is just a tooling, whether use kubectl or not isn't related with the user identity..

Reply:
and i renamed `system-low` to `workload-high`

Reply:
It is true that kubectl is not directly identified in the request attributes but the complement is nearly true.  After logically higher precedence levels have identified other stuff, what is left is workload.  That mainly consists of requests from controllers --- running in-cluster using a service account --- and requests from users using kubectl.  So the latter can be identified by subtracting out the former.  That is the approach you have been seeing from me since I had to start coping with the lack of negation in the matching language.